dbms/src/Flash/tests/gtest_collation.cpp

```diff
@@ -0,0 +1,436 @@
+                             {{limit_col, TiDB::TP::TypeString}},
+                             {toNullableVec<String>(limit_col, ColumnWithString{"col0-0", {}, "col0-2", "col0-3", {}, "col0-5", "col0-6", "col0-7"})});
+
+        /// For ExchangeSender
+        context.addExchangeRelationSchema(sender_name, {{"s1", TiDB::TP::TypeString}, {"s2", TiDB::TP::TypeString}, {"s3", TiDB::TP::TypeString}});
+    }
+
+    void setAndCheck(const String & table_name, const String & col_name, Int32 collation, const ColumnsWithTypeAndName & expect)
+    {
+        context.setCollation(collation);
+        auto request = context.scan(db_name, table_name).aggregation(MockAstVec{}, {col(col_name)}).project({col_name}).build(context);
+        std::cout << request->DebugString() << std::endl;
+        ASSERT_COLUMNS_EQ_UR(expect, executeStreams(request, 1));
+    }
+
+    void checkExecutorCollation(std::shared_ptr<tipb::DAGRequest> dag_request) const;
+    void checkScalarFunctionCollation(std::shared_ptr<tipb::DAGRequest> dag_request) const;
+    void addExpr(std::queue<const tipb::Expr *> & exprs, const tipb::Expr * const expr) const;
+
+    /// Prepare some names
+    const String db_name{"test_db"};
+    const String table_name{"collation_table"};
+    const String col_name{"col"};
+    const ColumnWithNullableString col{"china", "china", "china  ", "CHINA", "cHiNa ", "usa", "usa", "usa  ", "USA", "USA "};
+
+    const String chinese_table{"chinese"};
+    const String chinese_col_name{"col"};
+    const ColumnWithNullableString chinese_col{"北京", "北京  ", "北bei京", "北Bei京", "北bei京  ", "上海", "上海  ", "shanghai  ", "ShangHai", "ShangHai  "};
+
+    const String join_table{"join_table"};
+    const String topn_table{"topn_table"};
+    const String topn_col{"topn_col"};
+    const String proj_table{"proj_table"};
+    const std::vector<String> proj_col{"proj_col0", "proj_col1"};
+    const String limit_table{"limit_table"};
+    const String limit_col{"limit_col"};
+    const String sender_name{"sender"};
+
+    /// scalar functions whose collation must be set(Some more scalar functions may be added in the future)
+    std::set<int> scalar_func_need_collation{tipb::ScalarFuncSig::EQInt, tipb::ScalarFuncSig::NEInt, tipb::ScalarFuncSig::GTInt, tipb::ScalarFuncSig::LTInt};
+};
+
+/// Collect scalar functions
+void ExecutorCollation::addExpr(std::queue<const tipb::Expr *> & exprs, const tipb::Expr * const expr) const
+{
+    if (expr->tp() == tipb::ExprType::ScalarFunc) /// only add scalar function
+        exprs.push(expr);
+    int children_size = expr->children_size();
+
+    /// recursively add expression
+    for (int i = 0; i < children_size; ++i)
+        addExpr(exprs, &(expr->children(i)));
+}
+
+void ExecutorCollation::checkExecutorCollation(std::shared_ptr<tipb::DAGRequest> dag_request) const
+{
+    std::queue<tipb::Executor *> executors;
+    tipb::Executor * executor = dag_request->mutable_root_executor();
+    executors.push(executor);
```

Comment:
We can use https://github.com/pingcap/tiflash/blob/e289119502cb4f2643eb1810dffb4434a611b0a9/dbms/src/Flash/Statistics/traverseExecutors.h to visit DAGRequest

Reply:
> We can use https://github.com/pingcap/tiflash/blob/e289119502cb4f2643eb1810dffb4434a611b0a9/dbms/src/Flash/Statistics/traverseExecutors.h to visit DAGRequest

Done